File: src/test/scala/com/github/kayjamlang/tests/expressions/array/ArrayExpressionMoreGetTest.scala
package com.github.kayjamlang.tests.expressions.array

import com.github.kayjamlang.core.{KayJamLexer, KayJamParser}
import com.github.kayjamlang.core.expressions.{GetExpression, ValueExpression, VariableLinkExpression}
import org.junit.Assert._
import org.junit.{BeforeClass, Test}


object ArrayExpressionMoreGetTest {
  private var parser: KayJamParser = null

  @BeforeClass def prepare(): Unit = {
    parser = new KayJamParser(new KayJamLexer("test[2005][2006]"))
  }
}

class ArrayExpressionMoreGetTest {
  @Test
  @throws[Exception]
  def test(): Unit = {
    val expression = ArrayExpressionMoreGetTest.parser.readExpression
    assertNotNull(expression)
    assertSame(classOf[GetExpression], expression getClass)
    val arrayGet = expression.asInstanceOf[GetExpression]
    assertSame(classOf[GetExpression], arrayGet.root getClass)
    assertSame(classOf[ValueExpression], arrayGet.value getClass)
    val getExpression = arrayGet.root.asInstanceOf[GetExpression]
    assertSame(classOf[VariableLinkExpression], getExpression.root getClass)
    assertSame(classOf[ValueExpression], getExpression.value getClass)
    val constant = arrayGet.value.asInstanceOf[ValueExpression]
    assertEquals(2006, constant.value)
  }
}

File: src/test/java/com/github/kayjamlang/tests/expressions/array/ArrayExpressionMoreGetTest.java
package com.github.kayjamlang.tests.expressions.array;

import com.github.kayjamlang.core.expressions.Expression;
import com.github.kayjamlang.core.KayJamLexer;
import com.github.kayjamlang.core.KayJamParser;
import com.github.kayjamlang.core.expressions.ValueExpression;
import com.github.kayjamlang.core.expressions.GetExpression;
import com.github.kayjamlang.core.expressions.VariableLinkExpression;
import org.junit.BeforeClass;
import org.junit.Test;

import static org.junit.Assert.*;

public class ArrayExpressionMoreGetTest {

    private static KayJamParser parser;

    @BeforeClass
    public static void prepare(){
        parser = new KayJamParser(new KayJamLexer("test[2005][2006]"));
    }

    @Test
    public void test() throws Exception {
        Expression expression = parser.readExpression();

        assertNotNull(expression);
        assertSame(GetExpression.class, expression.getClass());

        GetExpression arrayGet = (GetExpression) expression;
        assertSame(GetExpression.class, arrayGet.root.getClass());
        assertSame(ValueExpression.class, arrayGet.value.getClass());

        GetExpression getExpression = (GetExpression) arrayGet.root;
        assertSame(VariableLinkExpression.class, getExpression.root.getClass());
        assertSame(ValueExpression.class, getExpression.value.getClass());

        ValueExpression constant = (ValueExpression) arrayGet.value;
        assertEquals(2006, constant.value);
    }
}